File: Data Structures & Algorithms/valid-tree/submission-4.cpp
class Solution {
public:
    bool validTree(int n, vector<vector<int>>& edges) {
        vector<vector<int>>graph(n,vector<int>(n,0));
        for(int i=0;i<edges.size();i++) {
            int fNode=edges[i][0];
            int lNode=edges[i][1];
            graph[fNode][lNode]=1;
            graph[lNode][fNode]=1;
        }

        vector<bool>visited(n,false);
        vector<int>parents(n,-1);   
        stack<int>cNodes;
        cNodes.push(0);
        while(!cNodes.empty()) {
            int cNode=cNodes.top();
            cNodes.pop();
            visited[cNode]=true;

            for(int i=0;i<n;i++) {
                if(graph[cNode][i]==1 && parents[cNode]!=i) {
                    if(visited[i]){
                        cout<<"Visited: "<<i<<" "<<"edge "<<cNode<<" "<<i<<endl;
                        cout<<"Parent of "<<i<<" "<<parents[i]<<endl;
                        return false;
                    }
                    cNodes.push(i);
                    parents[i]=cNode;
                }
            }
        }

        for(int node:visited) {
            if(!node)return false;
        }
        return true;

    }
};

File: Data Structures & Algorithms/count-connected-components/submission-4.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        tot_component=0
        visited=[False for _ in range(n)] 
        graph=[[] for _ in range(n)]
        for edge in edges:
            fNode=edge[0]
            lNode=edge[1]
            graph[fNode].append(lNode)
            graph[lNode].append(fNode)

        def bfs(node):
            nonlocal tot_component
            node_q=[]
            node_q.append(node)
            tot_component+=1
            while len(node_q)>0:
                cNode=node_q.pop()
                #visited[node]=True
                for node in graph[cNode]:
                    if visited[node]==False:
                        visited[node]=True
                        node_q.append(node)
        
        for i in range(n):
            if visited[i]==False:
                visited[i]=True
                bfs(i)
        
        return tot_component

        